### PR TITLE
feat: add tutorials and recaptcha-validations

### PR DESCRIPTION
**Pull request by Valliammai-SM**

### CODE REVIEW

File: server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt [hunk 1]
```diff
@@ -151,4 +196,15 @@ class UserService(
             )
         )
     }
+
+    fun updateTutorialLevel(userId: UUID, tutorialLevelResponseDto: TutorialLevelResponseDto) {
+        val (level) = tutorialLevelResponseDto
+        val user = userRepository.findById(userId).get()
+        userRepository.save(user.copy(tutorialLevel = level?.plus(1)))
+    }
+
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt]: why `updatetutorial` is a separate function in the `userservice`. Tutorial update is present with this common [/user(patch)](https://delta.github.io/codecharacter-server-2023/spec/index.html#/current-user/updateCurrentUser)  route. Update it over there

File: server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt [hunk 2]
```diff
@@ -133,6 +162,22 @@ class UserService(
         }
     }
 
+    fun verifyReCaptcha(reCaptchaResponse: String): Boolean {
+        val url =
+            "https://www.google.com/recaptcha/api/siteverify"
+        try {
+            val requestBody = ObjectMapper().writeValueAsString(mapOf("secret" to secretKey,"response" to reCaptchaResponse))
+            val client=HttpClient.newBuilder().build()
+            val request=HttpRequest.newBuilder().uri(URI.create(url)).POST(HttpRequest.BodyPublishers.ofString(requestBody)).build()
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt]:
```suggestion
            val request=HttpRequest.newBuilder().uri(URI.create(recaptchaUrl)).POST(HttpRequest.BodyPublishers.noBody()).build()
```
- `Request-body` should not be passed

File: server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt [hunk 3]
```diff
@@ -133,6 +162,22 @@ class UserService(
         }
     }
 
+    fun verifyReCaptcha(reCaptchaResponse: String): Boolean {
+        val url =
+            "https://www.google.com/recaptcha/api/siteverify"
+        try {
+            val requestBody = ObjectMapper().writeValueAsString(mapOf("secret" to secretKey,"response" to reCaptchaResponse))
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt]:
```suggestion
            val recaptchaUrl =             "https://www.google.com/recaptcha/api/siteverify?secret=${secretKey}&response=${reCaptchaResponse}"
```
- The `recaptcha-url` will be accepting the `secret` and `recaptchaCode` as `params` not as `request body`. [Check Here](https://developers.google.com/recaptcha/docs/verify)

File: server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt [hunk 4]
```diff
@@ -133,6 +162,22 @@ class UserService(
         }
     }
 
+    fun verifyReCaptcha(reCaptchaResponse: String): Boolean {
+        val url =
+            "https://www.google.com/recaptcha/api/siteverify"
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/user/UserService.kt]:
```suggestion
 
```

File: server/src/main/resources/application.example.yml [hunk 5]
```diff
@@ -39,5 +39,8 @@ cors:
 jwt:
   secret: 'example'
 
+user:
+  reCaptcha-key: your-reCaptcha-key
```
Comment (Ram-20062003) [inline server/src/main/resources/application.example.yml]: - `user.reCaptcha-key`->`environment.reCaptcha-key (yaml-format)
- Make corresponding changes in `application.docker.example.yml` also

**Review by Ram-20062003 — CHANGES_REQUESTED** [review]
## Changes Requested
- [x] Pull rebase your PR
- [x] Change `user.reCaptcha-key`->`environment.reCaptcha-key` (yaml format)
- [x] Make your corresponding changes in `application.docker.example.yml` also
- [x] `UpdateTutorial` is not proper. Fix it
- [x] Fix recaptcha verification
- [ ] Write `test` to handle the `tier-based-matching` customexception

_(Please make the following changes @Valliammai-SM . If you have any doubts let us know)_

**Review by Ram-20062003 — APPROVED** [review]

**Review by Ashwin125 — APPROVED** [review]

**Review by Ashwin125 — APPROVED** [review]

**Review by vigneshd332 — APPROVED** [review]

**Review by Ram-20062003 — APPROVED** [review]

**Pull request merged**